File: backend/utils/data_utils.py
import pandas as pd
from config import Config

def validate_csv_columns(df):
    """Validate that CSV contains expected columns"""
    missing_columns = [col for col in Config.EXPECTED_COLUMNS if col not in df.columns]
    return len(missing_columns) == 0, missing_columns

def get_basic_stats(df):
    """Get basic statistics from dataframe"""
    if df is None or df.empty:
        return None
    
    return {
        'total_rows': len(df),
        'unique_pharmacies': df['Pharmacy'].nunique(),
        'unique_clusters': df['Cluster'].nunique(),
        'unique_metrics': df['Metric'].nunique(),
        'date_range': {
            'start': df['Date'].min(),
            'end': df['Date'].max()
        } if 'Date' in df.columns else None
    }

def filter_data_by_date_range(df, start_date=None, end_date=None):
    """Filter dataframe by date range"""
    if df is None:
        return df
    
    filtered_df = df.copy()
    
    if start_date:
        filtered_df = filtered_df[filtered_df['Date'] >= start_date]
    
    if end_date:
        filtered_df = filtered_df[filtered_df['Date'] <= end_date]
    
    return filtered_df

def filter_data_by_pharmacies(df, pharmacies):
    """Filter dataframe by selected pharmacies"""
    if df is None or not pharmacies:
        return df
    
    return df[df['Pharmacy'].isin(pharmacies)]

def format_currency(value):
    """Format value as currency"""
    try:
        return f"£{value:,.0f}"
    except:
        return str(value) 

File: backend-js/utils/dataUtils.js
const config = require('../config');

function validateCsvColumns(data) {
  if (!data || data.length === 0) {
    return { isValid: false, missingColumns: config.EXPECTED_COLUMNS };
  }
  
  const firstRow = data[0];
  const foundColumns = Object.keys(firstRow);
  
  // Check for exact matches first
  const missingColumns = config.EXPECTED_COLUMNS.filter(col => !foundColumns.includes(col));
  
  return {
    isValid: missingColumns.length === 0,
    missingColumns
  };
}

function getBasicStats(data) {
  if (!data || data.length === 0) {
    return null;
  }
  
  const uniquePharmacies = new Set(data.map(row => row.Pharmacy));
  const uniqueClusters = new Set(data.map(row => row.Cluster));
  const uniqueMetrics = new Set(data.map(row => row.Metric));
  
  const dates = data.map(row => new Date(row.Date)).filter(date => !isNaN(date));
  const minDate = dates.length > 0 ? new Date(Math.min(...dates)) : null;
  const maxDate = dates.length > 0 ? new Date(Math.max(...dates)) : null;
  
  return {
    totalRows: data.length,
    uniquePharmacies: uniquePharmacies.size,
    uniqueClusters: uniqueClusters.size,
    uniqueMetrics: uniqueMetrics.size,
    dateRange: minDate && maxDate ? {
      start: minDate.toISOString().split('T')[0],
      end: maxDate.toISOString().split('T')[0]
    } : null
  };
}

function filterDataByDateRange(data, startDate = null, endDate = null) {
  if (!data) return data;
  
  return data.filter(row => {
    const rowDate = new Date(row.Date);
    if (isNaN(rowDate)) return false;
    
    if (startDate && rowDate < new Date(startDate)) return false;
    if (endDate && rowDate > new Date(endDate)) return false;
    
    return true;
  });
}

function filterDataByPharmacies(data, pharmacies) {
  if (!data || !pharmacies || pharmacies.length === 0) {
    return data;
  }
  
  return data.filter(row => pharmacies.includes(row.Pharmacy));
}

function formatCurrency(value) {
  try {
    const numValue = parseFloat(value);
    if (isNaN(numValue)) return String(value);
    return `£${numValue.toLocaleString('en-GB', { maximumFractionDigits: 0 })}`;
  } catch {
    return String(value);
  }
}

function cleanFiscalYear(fiscalYear) {
  if (!fiscalYear) return null;
  
  const cleaned = String(fiscalYear).replace(/FY/i, '');
  const numeric = parseInt(cleaned, 10);
  return isNaN(numeric) ? null : numeric;
}

module.exports = {
  validateCsvColumns,
  getBasicStats,
  filterDataByDateRange,
  filterDataByPharmacies,
  formatCurrency,
  cleanFiscalYear
}; 